File: 24. Swap Nodes in Pairs.py
# https://leetcode.com/problems/swap-nodes-in-pairs/

# Given a linked list, swap every two adjacent nodes and return its head. You must
# solve the problem without modifying the values in the list's nodes (i.e., only
# nodes themselves may be changed.)

################################################################################

# SLL swap -> foward traversal
# prev -> (curr -> next) -> ... becomes
# prev -> (next -> curr) -> ...

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next: return head
        
        dummy = ListNode()
        dummy.next = head
        
        # prev -> curr
        prev = dummy
        curr = head
        
        while curr and curr.next: # stop at curr = tail or tail.next
            # prev -> (curr -> next) -> ...
            next = curr.next 
            
            # want to do: prev -> (next -> curr) -> ...
            prev.next = next
            curr.next, next.next = next.next, curr

            prev, curr = curr, curr.next # move on
        return dummy.next

File: 24. Swap Nodes in Pairs.cpp
// https://leetcode.com/problems/swap-nodes-in-pairs/

// Given a linked list, swap every two adjacent nodes and return its head. You must
// solve the problem without modifying the values in the list's nodes (i.e., only
// nodes themselves may be changed.)

////////////////////////////////////////////////////////////////////////////////

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (!head or !head->next) return head;
        ListNode *dummy = new ListNode();
        dummy->next = head;
        ListNode *ptrPrev = dummy, *ptrCurr = head;
        // swap ptrCurr and ptrCurr->next
        while (ptrCurr and ptrCurr->next) {
            // ptrPrev -> ptrCurr -> ptrNext ->
            ListNode *ptrNext = ptrCurr->next;
            ptrPrev->next = ptrNext;
            
            // ptrPrev -> ptrNext -> ptrCurr ->
            ptrCurr->next = ptrNext->next;
            ptrNext->next = ptrCurr; 
            
            // move on
            ptrPrev = ptrCurr;
            ptrCurr = ptrCurr->next;
        }
        return dummy->next;
    }
};
